src/autoscaler/api/brokerserver/broker_handler.go

```diff
@@ -136,32 +148,80 @@ func (h *BrokerHandler) BindServiceInstance(w http.ResponseWriter, r *http.Reque
 		handlers.WriteJSONResponse(w, http.StatusBadRequest, models.ErrorResponse{
 			Code:    "Bad Request",
 			Message: "Malformed or missing mandatory data",
 		})
 		return
 	}
 
+	err = h.bindingdb.CreateServiceBinding(bindingId, instanceId, body.AppID)
+	if err != nil {
+		if err == db.ErrAlreadyExists {
+			h.logger.Error("failed to create binding: binding already exists", err, lager.Data{"appId": body.AppID})
+			handlers.WriteJSONResponse(w, http.StatusConflict, models.ErrorResponse{
+				Code:    "Conflict",
+				Message: "An autoscaler service instance is already bound to the application. Multiple bindings are not supported."})
+			return
+		}
+		h.logger.Error("failed to save binding", err, lager.Data{"appId": body.AppID, "bindingId": bindingId, "instanceId": instanceId})
+		handlers.WriteJSONResponse(w, http.StatusInternalServerError, models.ErrorResponse{
+			Code:    "Interal-Server-Error",
+			Message: "Error creating service binding"})
+		return
+	}
+	cred, err := custom_metrics_cred_helper.CreateCustomMetricsCredential(body.AppID, h.policydb, custom_metrics_cred_helper.MaxRetry)
+	if err != nil {
+		//revert binding creating
+		h.logger.Error("failed to create custom metrics credential", err, lager.Data{"appId": body.AppID})
+		err = h.bindingdb.DeleteServiceBindingByAppId(body.AppID)
+		if err != nil {
+			h.logger.Error("failed to revert binding due to failed to create custom metrics credential", err, lager.Data{"appId": body.AppID})
+		}
+		handlers.WriteJSONResponse(w, http.StatusInternalServerError, models.ErrorResponse{
+			Code:    "Interal-Server-Error",
+			Message: "Error creating service binding"})
+		return
+	}
 	if body.Policy == "" {
 		h.logger.Info("no policy json provided", lager.Data{})
 	} else {
 		errResults, valid := h.policyValidator.ValidatePolicy(body.Policy)
-
 		if !valid {
+			h.logger.Error("failed to validate policy", err, lager.Data{"appId": body.AppID, "policy": body.Policy})
+			//revert creating binding and custom metrics credential
+			err = custom_metrics_cred_helper.DeleteCustomMetricsCredential(body.AppID, h.policydb, custom_metrics_cred_helper.MaxRetry)
+			if err != nil {
+				h.logger.Error("failed to revert custom metrics credential due to failed to validate policy", err, lager.Data{"appId": body.AppID})
+			}
+			err = h.bindingdb.DeleteServiceBindingByAppId(body.AppID)
+			if err != nil {
+				h.logger.Error("failed to revert binding due to failed to validate policy", err, lager.Data{"appId": body.AppID})
+			}
 			handlers.WriteJSONResponse(w, http.StatusBadRequest, errResults)
 			return
 		}
-
 		policyGuid, err := uuid.NewV4()
 		if err != nil {
+			h.logger.Error("failed to create policy guid", err, lager.Data{"appId": body.AppID})
+			//revert creating binding and custom metrics credential
+			err = custom_metrics_cred_helper.DeleteCustomMetricsCredential(body.AppID, h.policydb, custom_metrics_cred_helper.MaxRetry)
```

Comment:
Please do the policy validation and uuid generation prior the service binding and credential generation, so that we can avoid the meaningless db rollback.